mailbox/tika/src/main/java/org/apache/james/mailbox/tika/CachingTextExtractor.java

```
@@ -0,0 +1,84 @@
+
+import org.apache.commons.codec.digest.DigestUtils;
+import org.apache.commons.io.IOUtils;
+import org.apache.james.mailbox.extractor.ParsedContent;
+import org.apache.james.mailbox.extractor.TextExtractor;
+
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.Weigher;
+import com.google.common.util.concurrent.UncheckedExecutionException;
+
+
+public class CachingTextExtractor implements TextExtractor {
+    private final TextExtractor underlying;
+    private final Cache<String, ParsedContent> cache;
+
+    public CachingTextExtractor(TextExtractor underlying, Duration cacheEvictionPeriod, Long cacheWeightInBytes) {
+        this.underlying = underlying;
+
+        Weigher<String, ParsedContent> weigher = (key, parsedContent) -> getSize(parsedContent);
+        this.cache = CacheBuilder.<String, String>newBuilder()
+            .expireAfterAccess(cacheEvictionPeriod.toMillis(), TimeUnit.MILLISECONDS)
+            .maximumWeight(cacheWeightInBytes)
+            .weigher(weigher)
+            .build();
+    }
+
+    private int getSize(ParsedContent parsedContent) {
+        return parsedContent.getTextualContent()
+            .map(String::length)
+            .map(this::utf16LengthToBytesCount)
+            .orElse(0);
+    }
+
+    private int utf16LengthToBytesCount(Integer value) {
+        return value * 2;
+    }
+
+    @Override
+    public ParsedContent extractContent(InputStream inputStream, String contentType) throws Exception {
+        byte[] bytes = IOUtils.toByteArray(inputStream);
+        String key = DigestUtils.sha256Hex(bytes);
+        try {
+            return cache.get(key,
+                () -> underlying.extractContent(new ByteArrayInputStream(bytes), contentType));
+        } catch (UncheckedExecutionException | ExecutionException e) {
+            return unwrap(e);
+        }
+    }
+
+    public ParsedContent unwrap(Exception e) throws Exception {
```

you can return the Exception and let the caller throw, I would reduce boilerplate (like saying you return something you don't actually return)